cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java

```diff
@@ -542,50 +567,59 @@ public void increment(Increment increment) {
     table.increment(increment);
   }
 
+  @ReadOnly
   @Override
   public Scanner scan(@Nullable byte[] startRow, @Nullable byte[] stopRow) {
     return table.scan(startRow, stopRow);
   }
 
+  @ReadOnly
   @Override
   public Scanner scan(Scan scan) {
     return table.scan(scan);
   }
 
   /* BatchReadable implementation */
 
+  @ReadOnly
   @Override
   public List<Split> getSplits(int numSplits, byte[] start, byte[] stop) {
     return table.getSplits(numSplits, start, stop);
   }
 
+  @NoAccess
   @Override
   public Type getRecordType() {
     return table.getRecordType();
   }
 
+  @WriteOnly
   @Override
   public void write(StructuredRecord structuredRecord) throws IOException {
     table.write(structuredRecord);
   }
 
+  @ReadOnly
   @Override
   public List<Split> getSplits() {
     return table.getSplits();
   }
 
+  @ReadOnly
   @Override
   public RecordScanner<StructuredRecord> createSplitRecordScanner(Split split) {
     return table.createSplitRecordScanner(split);
   }
 
+  @ReadOnly
   @Override
   public SplitReader<byte[], Row> createSplitReader(Split split) {
     return table.createSplitReader(split);
   }
 
   /* BatchWritable implementation */
 
+  @WriteOnly
```

Comment:
With the new lineage handling, a `WRITE` lineage will be record, while a READ and WRITE authorization will be enforced and audit log being published when the actual data operations happen